File: cpp/math/RabinMiller.cpp
#include <iostream>
#include <vector>
using namespace std;

class RabinMiller {
public:
    RabinMiller() {
        int aux[10] = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29};
        primes.assign(aux, aux + 10);
    }

    bool isPrime(long long num) {
        if (primes.end() != find(primes.begin(), primes.end(), num)) {
            return true;
        }
        if (num < 2 || !(num & 1)) {
            return false;
        }

        long long base = num - 1;
        int index = 0;
        while (!(base & 1)) {
            base >>= 1;
            ++index;
        }

        for (int value : primes) {
            if (check(value, num, base, index)) {
                return false;
            }
        }
        return true;
    }

private:
    vector<int> primes;

    bool check(long long factor, long long num, long long base, int index) {
        long long pivot = powerMod(factor, base, num);
        if (1 == pivot) {
            return false;
        }

        while (index) {
            long long value = multiplyMod(pivot, pivot, num);
            if (1 == value && 1 != pivot && pivot != num - 1) {
                return true;
            }
            pivot = value;
            --index;
        }
        return 1 != pivot;
    }

    long long powerMod(long long base, long long index, long long mod) {
        base %= mod;
        if (1 == index) {
            return base;
        }

        long long aux = base;
        long long result = 1;
        while (index) {
            if (index & 1) {
                result = multiplyMod(result, aux, mod);
            }
            aux = multiplyMod(aux, aux, mod);
            index >>= 1;
        }
        return result;
    }

    long long multiplyMod(long long i, long long j, long long mod) {
        i %= mod;
        j %= mod;
        long long result = 0;
        while (j) {
            if (j & 1) {
                result = (result + i) % mod;
            }
            i <<= 1;
            if (i >= mod) {
                i %= mod;
            }
            j >>= 1;
        }
        return result;
    }
};

File: py/math/RabinMiller.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class RabinMiller:
    def __init__(self):
        self.__primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]

    def isPrime(self, num: int) -> bool:
        if num in self.__primes:
            return True
        if num < 2 or not num & 1:
            return False

        base, index = num - 1, 0
        while not base & 1:
            base >>= 1
            index += 1

        for value in self.__primes:
            if self.__check(value, num, base, index):
                return False
        return True

    def __check(self, factor: int, num: int, base: int, index: int) -> bool:
        pivot = self.__powerMod(factor, base, num)
        if 1 == pivot:
            return False

        while index:
            value = self.__multiplyMod(pivot, pivot, num)
            if 1 == value and 1 != pivot and pivot != num - 1:
                return True
            pivot = value
            index -= 1
        return 1 != pivot

    def __powerMod(self, base: int, index: int, mod: int) -> int:
        base %= mod
        if 1 == index:
            return base

        aux, result = base, 1
        while index:
            if index & 1:
                result = self.__multiplyMod(result, aux, mod)
            aux = self.__multiplyMod(aux, aux, mod)
            index >>= 1
        return result

    def __multiplyMod(self, i: int, j: int, mod: int) -> int:
        i %= mod
        j %= mod
        result = 0
        while j:
            if j & 1:
                result = (result + i) % mod
            i <<= 1
            if i >= mod:
                i %= mod
            j >>= 1
        return result
